app/demo-loop/KernelUtils.i.hh

```diff
@@ -32,76 +31,133 @@ CELER_FUNCTION void calc_step_limits(const GeoTrackView&      geo,
     }
+    phys.step_length(step);
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Propagate up to the step length or next boundary.
+ * Propagate up to the step length or next boundary, calculate the energy loss
+ * over the step, and select the model for the discrete interaction.
  */
-CELER_FUNCTION real_type propagate(GeoTrackView&           geo,
-                                   const PhysicsTrackView& phys)
+template<class Rng>
+CELER_FUNCTION void move_and_select_model(GeoTrackView&      geo,
+                                          ParticleTrackView& particle,
+                                          PhysicsTrackView&  phys,
+                                          SimTrackView&      sim,
+                                          Rng&               rng,
+                                          real_type*         edep,
+                                          Interaction*       result)
 {
     // Actual distance, limited by along-step length or geometry
     real_type step = phys.step_length();
     if (step > 0)
     {
+        // Store the current volume
+        auto pre_step_volume = geo.volume_id();
+
         // Propagate up to the step length or next boundary
         LinearPropagator propagate(&geo);
         auto             geo_step = propagate(step);
         step                      = geo_step.distance;
-        // TODO: check whether the volume/material have changed
+
+        // Particle entered a new volume before reaching the interaction point
+        if (geo_step.volume != pre_step_volume)
+        {
+            result->energy    = particle.energy();
+            result->direction = geo.dir();
+            result->action    = Action::entered_volume;
+        }
+    }
+    phys.step_length(phys.step_length() - step);
+
+    // Kill the track if it's outside the valid geometry region
+    if (geo.is_outside())
+    {
+        result->action = Action::escaped;
+        sim.alive(false);
+    }
+
+    // Calculate energy loss over the step length
+    auto eloss = calc_energy_loss(particle, phys, step);
+    *edep += eloss.value();
+    particle.energy(
+        ParticleTrackView::Energy{particle.energy().value() - eloss.value()});
+
+    // Reduce the remaining mean free path
+    real_type mfp = phys.interaction_mfp() - step * phys.macro_xs();
+    phys.interaction_mfp(soft_zero(mfp) ? 0 : mfp);
+
+    ModelId model{};
+    if (phys.interaction_mfp() <= 0)
+    {
+        // Reached the interaction point: sample the process and determine the
+        // corresponding model
+        auto ppid_mid = select_process_and_model(particle, phys, rng);
+        model         = ppid_mid.model;
     }
-    return step;
+    phys.model_id(model);
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Select the model for the discrete interaction.
+ * Apply secondary cutoffs and process interaction change.
```

Comment:
No, it's not guaranteed to be a post-interaction track. If it reached a boundary, it will have `interaction.action == Action::entered_volume` and `action_unchanged(...)` will return true. Looking at this again, I think I need to fix the `Interaction` for boundary-crossing tracks -- as-is it seems like the energy deposition could be nonzero.

Reply:
Yeah, we can definitely have energy deposition (along-step deposition) when hitting a boundary.